File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        n = len(nums)
        for i in range(n - 2):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            l = i + 1
            r = n - 1
            while l < r:
                three_sum = nums[i] + nums[l] + nums[r]
                if three_sum == 0:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l + 1] == nums[l]:
                        l += 1
                    l += 1
                    while l < r and nums[r - 1] == nums[r]:
                        r -= 1
                    r -= 1
                elif three_sum < 0:
                    l += 1
                else:
                    r -= 1
        return res

File: Data Structures & Algorithms/single-threaded-cpu/submission-2.java
class Solution {
    public int[] getOrder(int[][] tasks) {
        int[][] sortedTasks = new int[tasks.length][3];
        for (int i = 0; i < tasks.length; i++) {
            sortedTasks[i] = new int[] {tasks[i][0], tasks[i][1], i};
        }

        Arrays.sort(sortedTasks, (t1, t2) -> t1[0] - t2[0]);

        long cycle = 0;
        int[] res = new int[tasks.length];
        int currRes = 0;
        int currSortedTask = 0;
        PriorityQueue<int[]> pq2 =
            new PriorityQueue<>((t1, t2) -> t1[0] == t2[0] ? t1[1] - t2[1] : t1[0] - t2[0]);
        
        while (currSortedTask < tasks.length || !pq2.isEmpty()) {
            if (pq2.isEmpty() && cycle < sortedTasks[currSortedTask][0]) {
                cycle = sortedTasks[currSortedTask][0];
            }

            while (currSortedTask < tasks.length && sortedTasks[currSortedTask][0] <= cycle) {
                int[] task = sortedTasks[currSortedTask++];
                pq2.add(
                    new int[] {
                        task[1], task[2]
                    }
                );
            }

            int[] task = pq2.poll();
            res[currRes++] = task[1];
            cycle += task[0];
        }

        return res;
    }
}